port/cpl_minizip_zip.cpp

```diff
@@ -2141,6 +2272,297 @@ CPLErr CPLCloseFileInZip(void *hZip)
+
+    CPLZip *psZip = static_cast<CPLZip *>(hZip);
+    zip64_internal *zi = reinterpret_cast<zip64_internal *>(psZip->hZip);
+
+    std::unique_ptr<VSIVirtualHandle> poFileHandleAutoClose;
+    if (fpInput == nullptr)
+    {
+        fpInput = VSIFOpenL(pszInputFilename, "rb");
+        if (fpInput == nullptr)
+            return CE_Failure;
+        poFileHandleAutoClose.reset(
+            reinterpret_cast<VSIVirtualHandle *>(fpInput));
+    }
+
+    VSIFSeekL(fpInput, 0, SEEK_END);
+    const auto nUncompressedSize = VSIFTellL(fpInput);
+    VSIFSeekL(fpInput, 0, SEEK_SET);
+
+    CPLStringList aosNewsOptions(papszOptions);
+    bool bSeekOptimized = false;
+    const char *pszSOZIP =
+        CSLFetchNameValueDef(papszOptions, "SOZIP_ENABLED",
+                             CPLGetConfigOption("CPL_SOZIP_ENABLED", "AUTO"));
+
+    const char *pszChunkSize = CSLFetchNameValueDef(
+        papszOptions, "SOZIP_CHUNK_SIZE",
+        CPLGetConfigOption("CPL_VSIL_DEFLATE_CHUNK_SIZE", nullptr));
+    const bool bChunkSizeSpecified = pszChunkSize != nullptr;
+    if (pszChunkSize == nullptr)
+        pszChunkSize = "1024K";
+    unsigned nChunkSize = static_cast<unsigned>(atoi(pszChunkSize));
+    if (strchr(pszChunkSize, 'K'))
+        nChunkSize *= 1024;
+    else if (strchr(pszChunkSize, 'M'))
+        nChunkSize *= 1024 * 1024;
+    nChunkSize =
+        std::max(static_cast<unsigned>(1),
+                 std::min(static_cast<unsigned>(UINT_MAX), nChunkSize));
+
+    const char *pszMinFileSize = CSLFetchNameValueDef(
+        papszOptions, "SOZIP_MIN_FILE_SIZE",
+        CPLGetConfigOption("CPL_SOZIP_MIN_FILE_SIZE", "1M"));
+    uint64_t nSOZipMinFileSize = std::strtoull(pszMinFileSize, nullptr, 10);
+    if (strchr(pszMinFileSize, 'K'))
+        nSOZipMinFileSize *= 1024;
+    else if (strchr(pszMinFileSize, 'M'))
+        nSOZipMinFileSize *= 1024 * 1024;
+    else if (strchr(pszMinFileSize, 'G'))
+        nSOZipMinFileSize *= 1024 * 1024 * 1024;
+
+    std::vector<uint8_t> sozip_index;
+    uint64_t nExpectedIndexSize = 0;
+    constexpr unsigned nDefaultSOZipChunkSize = 32 * 1024;
+    constexpr size_t nOffsetSize = 8;
+    if (((EQUAL(pszSOZIP, "AUTO") && nUncompressedSize > nSOZipMinFileSize) ||
+         (!EQUAL(pszSOZIP, "AUTO") && CPLTestBool(pszSOZIP))) &&
+        ((bChunkSizeSpecified &&
+          nUncompressedSize > static_cast<unsigned>(nChunkSize)) ||
+         (!bChunkSizeSpecified && nUncompressedSize > nDefaultSOZipChunkSize)))
+    {
+        if (!bChunkSizeSpecified)
+            nChunkSize = nDefaultSOZipChunkSize;
+
+        bSeekOptimized = true;
+
+        aosNewsOptions.SetNameValue(
+            "UNCOMPRESSED_SIZE", CPLSPrintf(CPL_FRMT_GUIB, nUncompressedSize));
+
+        zi->nOffsetSize = nOffsetSize;
+        nExpectedIndexSize =
+            32 + ((nUncompressedSize - 1) / nChunkSize) * nOffsetSize;
+        if (nExpectedIndexSize >
+            static_cast<uint64_t>(std::numeric_limits<int>::max()))
```

Comment:
I'm afraid thousands of places in GDAL assume that int == int32_t. Here, this is a loose check. We check that the index size is not excessively big to avoid a crazy memory allocation. uint16_t is definitely too small.
I'll probably keep that code as it, because if there is a platform where int != int32_t, if it is one where int < int32_t, then GDAL is probably totally broken, and checking nExpectedIndexSize against such small int_max value would be quite reasonable. And for an hypothetical platform where int would be > int32_t then hard to tell if the check is probably useless, but later we catch a memory allocation error.